### ISSUE TITLE
Hook for pytorch

### ISSUE THREAD

**Issue by AceCoooool** [pyinstaller/pyinstaller#2666]
`pyinstaller test.py` is completed successfully. However, when run this exe file, there is the error like that:
```
File "/home/ace/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/PyInstaller/loader/pyimod03_importers.py", line 389, in load_module
    exec(bytecode, module.__dict__)
  File "site-packages/torch/__init__.py", line 302, in <module>
  File "site-packages/torch/__init__.py", line 297, in manager_path
RuntimeError: Unable to find torch_shm_manager at /home/ace/Developer/temp/dist/test/torch/lib/torch_shm_manager
Failed to execute script test
```

This is maily caused by `path`，torch_shm_manager is in the file **/home/ace/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/torch/lib/torch_shm_manager**. (this is because pyinstaller change the __file__  to the current file: **/home/ace/Developer/temp/dist/test**)
How can I just let my **torch package** not change the __file__ ? (Even I can not use this exe file in other computers)

Could anyone help me? thank you.

**Comment by AceCoooool** [comment]
I using a very silly way: add the torch package to the `/home/ace/Developer/dist/test` . That's work.

**Comment by iamtodor** [comment]
It takes me a while to figure out the solution from @AceCoooool 
Here is the code
```
cnn2@cnn2:~$ cd ~/.local/lib/python3.6/site-packages
```
I assume your torch is placed right there
```
cnn2@cnn2:~/.local/lib/python3.6/site-packages$ cp -a torch/. your/dist/location
```
It copies all content from torch to your desired location.

Basically, I think it's a bug that should be fixed @htgoebel

**Comment by htgoebel** [comment]
@iamtodor This way work, but it is not what I'd call a solution. A solution would be to provide a hook, so others will not step into these issue.

**Comment by htgoebel** [comment]
@iamtodor You are very welcome to provide a pull-request: https://pyinstaller.readthedocs.io/en/stable/hooks.html

**Comment by xiexupang** [comment]
Have you solve the problem?
Now I meet the similar problem. My test.py only several lines code:
import torch
a = torch.zeros(2, 3)
print(a)

pyinstaller test.py is completed successfully. However, when run this exe file, there is the error like that
Failed to execute script test.  I did as @AceCoooool  did. But it can't work. there is the error like that
Failed to execute script test.

**Comment by Vishwaatpurdue** [comment]
Hi,
I have a similar issue. I used the hook mentioned here to load the torch dependencies, but the exe is not working in other windows machines (I built the exe for windows).

**Comment by KnightOfTheMoonlight** [comment]
You may try reinstall pytorch. That works for me.

**Comment by MarcinMisiurewicz** [comment]
I've run into similar problem, but none of the workarounds work. What is strange, my .exe file works when run from anaconda environment (any env: with or without torch installed), but when I try to run it from cmd outside of anaconda it gives me error:
`Traceback (most recent call last):
  File "verify.py", line 1, in <module>
  File "C:\ProgramData\Anaconda3\envs\py367\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
    exec(bytecode, module.__dict__)
  File "site-packages\torch\__init__.py", line 79, in <module>
ImportError: DLL load failed: The specified module could not be found.
[11108] Failed to execute script verify`

line 1 in verify.py is `import torch`
Do you have any idea what causes it? Any hints on where to look for it?

Win 10
Anaconda with python 3.6.7 (on 3.7 also fails)
PyInstaller 3.4
PyTorch 1.1.0

**Comment by raviv** [comment]
@MarcinMisiurewicz Same here.

**Comment by MarcinMisiurewicz** [comment]
Wow, thanks, I'll try it.

**Comment by MarcinMisiurewicz** [comment]
@raviv Wow, it works! I'm wondering what is the source of this problem, is it pytorch for windows, or pyinstaller?

Additional note: when I'm using plain Python (without conda), with pytorch installed from pip, it also crushes, exactly like above. So it seems that the workaround is to use conda, but install pytorch with pip (???). I don't know, it is some dark wizardry here...

**Comment by raviv** [comment]
@MarcinMisiurewicz glad it worked for you.
It seems that pyinstaller doesn't pick up the dependency of pytorch from conda, on cudatoolkit10.

**Comment by MarcinMisiurewicz** [comment]
when I've checked the dependency walker it think it was both: cudatoolkit and Python37.dll. I'll check it

**Comment by raviv** [comment]
@MarcinMisiurewicz After removing pytorch from conda and installing it using pip (in coda env) I can now build an exe that runs on a clean windows machine, with CUDA toolkit and driver installed of course.

**Comment by parth52** [comment]
> @raviv glad it worked for you.
> It seems that pyinstaller doesn't pick up the dependency of pytorch from conda, on cudatoolkit10.

i made a simple application to calculate speed between cpu and gpu and its works fine in my anaconda environment but whenever i am trying to make .exe of that particular file which raise an  error that cudatoolkit is not found do conda install cudatoolkit

my code as below temp.py file 

from numba import jit,cuda
import numpy as np 
# to measure exec time 
from timeit import default_timer as timer    
  
# normal function to run on cpu 
def func(a):                                 
    for i in range(10000000): 
        a[i]+= 1      
  
# function optimized to run on gpu  
@jit(target ="cuda")                          
def func2(a): 
    for i in range(10000000): 
        a[i]+= 1
if __name__=="__main__": 
    n = 10000000                            
    a = np.ones(n, dtype = np.float64) 
    b = np.ones(n, dtype = np.float32) 
      
    start = timer() 
    func2(a) 
    print("with GPU:", timer()-start) 
       
      
    start = timer() 
    func(a) 
    print("without GPU:", timer()-start)  
    ll=input()
    print(ll)

and error which i got as below while i am running exe file:-

 cudatoolkit is not found

whereas i installed cudatoolkit=9.0 in my conda environment.

and if i try to make .exe using either conda or cmd its give me an  same error

**Comment by jurneo** [comment]
> You may try reinstall pytorch. That works for me.

this is also a solution for me.

**Comment by bwoodsend** [comment]
This was fixed in #3657.

**Comment by bmahlbrand** [comment]
> This was fixed in #3657.

definitely still a thing in 4.4

**Comment by bwoodsend** [comment]
Open a new issue if you want us to be able do anything about it. I imagine that pytorch has changed a bit since 3 and a half years ago.
